crates/zksync/compiler/src/zksolc/mod.rs

```diff
@@ -120,76 +180,136 @@ impl DualCompiledContracts {
-                    // bytes. However, we should check and
-                    // handle errors in case an Unlinked BytecodeObject gets
-                    // here somehow
-                    let bytecode_vec = bytecode.object.clone().into_bytes().unwrap().to_vec();
-                    let mut factory_deps_vec: Vec<Vec<u8>> = factory_deps_map
-                        .keys()
-                        .map(|factory_hash| zksolc_all_bytecodes.get(factory_hash).unwrap())
-                        .cloned()
-                        .collect();
-
-                    factory_deps_vec.push(bytecode_vec.clone());
-
-                    dual_compiled_contracts.push(DualCompiledContract {
-                        name: contract_name,
-                        zk_bytecode_hash: H256::from_str(hash).unwrap(),
-                        zk_deployed_bytecode: bytecode_vec,
-                        zk_factory_deps: factory_deps_vec,
-                        evm_bytecode_hash: keccak256(solc_deployed_bytecode),
-                        evm_bytecode: solc_bytecode.to_vec(),
-                        evm_deployed_bytecode: solc_deployed_bytecode.to_vec(),
-                    });
-                } else {
-                    tracing::error!("matching solc artifact not found for {contract_file:?}");
-                }
+                // TODO: we can do this because no bytecode object could be unlinked
+                // at this stage for zksolc, and BytecodeObject as ref will get the bytecode
+                // bytes. However, we should check and
+                // handle errors in case an Unlinked BytecodeObject gets
+                // here somehow
+                let bytecode_vec = bytecode.object.clone().into_bytes().unwrap().to_vec();
+                let mut factory_deps_vec: Vec<Vec<u8>> = factory_deps_map
+                    .keys()
+                    .map(|factory_hash| zksolc_all_bytecodes.get(factory_hash).unwrap())
+                    .cloned()
+                    .collect();
+
+                factory_deps_vec.push(bytecode_vec.clone());
+
+                let zk_contract = Some(ZkContract {
+                    bytecode_hash: H256::from_str(hash).unwrap(),
+                    deployed_bytecode: bytecode_vec,
+                    factory_deps: factory_deps_vec,
+                });
+
+                this.push(DualCompiledContract {
+                    name: contract_name,
+                    path: Some(contract_file),
+                    zk: zk_contract,
+                    evm: None,
+                });
             }
         }
 
-        Self { contracts: dual_compiled_contracts }
+        this
+    }
+
+    /// Merge 2 [`DualCompiledContracts`] instances
+    ///
+    /// When a contract in the collections have the same components, the `other`'s is used
+    fn _merge(this: DualCompiledContracts, mut other: DualCompiledContracts) -> Self {
```

Comment:
1) zkSolc and solc are not exclusive to each other, they can both be present
2) it's mostly to avoid having to duplicate code in the `new_dual` constructor